smbclient/shutil.py

```diff
@@ -0,0 +1,306 @@
+        raise ValueError(
+            "Server side copy can only occur on the same drive, '%s' must be the same as the dst root '%s'" % (
+                src_drive, dst_drive))
+
+    try:
+        stat(norm_dst, **kwargs)
+    except SMBOSError as err:
+        if err.errno != errno.ENOENT:
+            raise
+    else:
+        raise ValueError("Target %s already exists" % norm_dst)
+
+    with SMBRawIO(norm_src, mode='r', desired_access=FilePipePrinterAccessMask.GENERIC_READ,
+                  create_options=(CreateOptions.FILE_NON_DIRECTORY_FILE), share_access='r',
+                  **kwargs) as raw_src:
+
+        with SMBFileTransaction(raw_src) as transaction_src:
+            ioctl_request(transaction_src, CtlCode.FSCTL_SRV_REQUEST_RESUME_KEY,
+                          flags=IOCTLFlags.SMB2_0_IOCTL_IS_FSCTL,
+                          output_size=32)
+
+        val_resp = SMB2SrvRequestResumeKey()
+        val_resp.unpack(transaction_src.results[0])
+
+        chunks = _get_srv_copy_chunks(transaction_src)
+
+        with SMBRawIO(norm_dst, mode='x', desired_access=FilePipePrinterAccessMask.GENERIC_WRITE,
+                      share_access='r',
+                      create_options=(CreateOptions.FILE_NON_DIRECTORY_FILE), **kwargs) as raw_dst:
+
+            with SMBFileTransaction(raw_dst) as transaction_dst:
+                for batch in _batches(chunks, 16):
+                    copychunkcopy_struct = SMB2SrvCopyChunkCopy()
+                    copychunkcopy_struct['source_key'] = val_resp['resume_key'].get_value()
+                    copychunkcopy_struct['chunks'] = batch
+
+                    ioctl_request(transaction_dst, CtlCode.FSCTL_SRV_COPYCHUNK_WRITE,
+                                  flags=IOCTLFlags.SMB2_0_IOCTL_IS_FSCTL,
+                                  output_size=32, buffer=copychunkcopy_struct)
+
+            for result in transaction_dst.results:
+                copychunk_response = SMB2SrvCopyChunkResponse()
+                copychunk_response.unpack(result)
+                if copychunk_response['chunks_written'].get_value() < 1:
+                    raise SMBOSError('Could not copy chunks in server side copy', filename=norm_dst)
+
+
+def _get_srv_copy_chunks(transaction):
+    chunks = []
+    offset = 0
+
+    while offset < transaction.raw.fd.end_of_file:
+        copychunk_struct = SMB2SrvCopyChunk()
+        copychunk_struct['source_offset'] = offset
+        copychunk_struct['target_offset'] = offset
+        if offset + CHUNK_SIZE < transaction.raw.fd.end_of_file:
+            copychunk_struct['length'] = CHUNK_SIZE
+        else:
+            copychunk_struct['length'] = transaction.raw.fd.end_of_file - offset
+
+        chunks.append(copychunk_struct)
+        offset += CHUNK_SIZE
+
+    return chunks
+
+
+def _batches(lst, n):
+    for i in range(0, len(lst), n):
+        yield lst[i:i + n]
+
+
+def _samefile(src, dst):
+    return normcase(abspath(src)) == normcase(abspath(dst))
```

Comment:
If you want a proper samefile then `smbclient.path` has a function that actually compares the file id on the host but it is only really designed for remote -> remote comparisons.

Reply:
I have extended samefile a bit to deal with certain scenarios.